gensim/similarities/levenshtein.py

```diff
@@ -8,146 +8,101 @@
+    beta : float
+        The exponential factor beta defined by [charletetal17]_.
+    index : :class:`lexpy.dawg.DAWG`
+        The DAWG nearest-neighbor search index.
+    max_distance : int
+        The maximum Levenshtein distance of the most similar terms.
+        Keeping this value below 3 has a significant impact on the
+        retrieval performance.
 
     See Also
     --------
-    :func:`gensim.similarities.levenshtein.levsim`
-        The Levenshtein similarity.
+    :class:`~gensim.similarities.termsim.WordEmbeddingSimilarityIndex`
+        Retrieve most similar terms for a given term using the cosine
+        similarity over word embeddings.
     :class:`~gensim.similarities.termsim.SparseTermSimilarityMatrix`
         Build a term similarity matrix and compute the Soft Cosine Measure.
 
+    References
+    ----------
+    The Levenshtein similarity, as a function of Levenshtein distance, was
+    defined in the context of term similarity by [charletetal17]_.
+
+    .. [charletetal17] Delphine Charlet and Geraldine Damnati, "SimBow at SemEval-2017 Task 3:
+       Soft-Cosine Semantic Similarity between Questions for Community Question Answering", 2017,
+       https://www.aclweb.org/anthology/S17-2051/.
+
     """
-    def __init__(self, dictionary, alpha=1.8, beta=5.0, threshold=0.0):
+    def __init__(self, dictionary, alpha=1.8, beta=5.0, max_distance=1):
         self.dictionary = dictionary
         self.alpha = alpha
         self.beta = beta
-        self.threshold = threshold
+        self.max_distance = max_distance
+
+        self.index = FastSS(self.max_distance)
+        for term in self.dictionary.values():
+            self.index.add(term)
+
         super(LevenshteinSimilarityIndex, self).__init__()
 
+    def _levsim(self, t1, t2, distance):
+        max_lengths = max(len(t1), len(t2)) or 1
+        similarity = self.alpha * (1.0 - distance * 1.0 / max_lengths)**self.beta
+        return similarity
+
     def most_similar(self, t1, topn=10):
-        similarities = (
-            (levsim(t1, t2, self.alpha, self.beta, self.threshold), t2)
-            for t2 in self.dictionary.values()
-            if t1 != t2
-        )
-        most_similar = (
-            (t2, similarity)
-            for (similarity, t2) in sorted(similarities, reverse=True)
-            if similarity > 0
-        )
-        return itertools.islice(most_similar, int(topn))
+        most_similar = []
+        if self.max_distance > 0:
+            effective_topn = topn + 1 if t1 in self.dictionary.token2id else topn
+            effective_topn = min(len(self.dictionary), effective_topn)
+            for distance in range(1, self.max_distance + 1):
+                terms = self.index.query(t1, distance)[distance]
+                for t2 in terms:
+                    if t1 == t2:
+                        continue
+                    similarity = self._levsim(t1, t2, distance)
+                    most_similar.append((t2, similarity))
```

Comment:
We are iteratively adding nodes with distance 1, then nodes with distance 2, and so on.

Reply:
Oh, you are right. We are iteratively adding words at distance 1, then distance 2, then distance 3, etc. However, this does not prevent us from getting similarity zero. :-)

Reply:
This algo scales badly with `k` – I wouldn't recommend anything higher than `k=2`.

Reply:
Isn't it better to do the check here, at this point of `append`, before the `break`? Otherwise you may end up with fewer than `effective_topn` results, after filtering out the zeros.

Reply:
Fixed in 6c2d033 and f651e639.

Reply:
> Otherwise you may end up with fewer than effective_topn results, after filtering out the zeros.

If we eagerly filter the zeros, we may never retrieve `topn` results and we will keep backing off.

Reply:
Both fixes are off IMO (see my comment above about `effective_topn`).

Reply:
If some of the similarities are zero, we will never get any non-zero similarities with futher back-offs.

Reply:
Therefore, it makes no sense to filter the zeroes eagerly and try to increase the distance.

Reply:
Isn't backing off until we retrieve `topn` results (or reach `max_distance`) a feature? I don't get it.

Reply:
Sorry, you are right. We actually can get zeros and still receive further non-zeros.

Reply:
Let me push what I think should happen. Hopefully you're not editing too, to prevent conflicts.

Reply:
Fixed in 80cdb7e and c91bda5.

Reply:
> But isn't backing off until we retrieve `topn` results (or reach `max_distance`) a feature? I don't get it.

You are right. I just assumed that the similarities decrease as the distances increase, and they do to some extent, but it's not strictly increasing. We may get non-zeros after zeros.

Reply:
OK. Given this behaviour, isn't it also possible we'll miss some results because they have high distance (discarded) which nevertheless results in high similarity (should be kept)?

In other words, are we sure the algo gives correct results when we use a different metric for the pruning? (prune by low distance but rank by high similarity)

I have no intuition how "Levenshtein similarity" behaves, so it's hard for me to reason about it.